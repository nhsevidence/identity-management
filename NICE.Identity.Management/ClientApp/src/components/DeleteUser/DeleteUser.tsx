import React, { Component } from "react";

import { Endpoints } from "../../data/endpoints";

type DeleteUserProps = {
	id: Number;
	onDeleteUser: Function;
};

export class DeleteUser extends Component<DeleteUserProps> {
	fetchDeleteData = async (url: string, id: Number) => {
		const response = await fetch(url, {
			method: "DELETE",
			headers: { "Content-Type": "application/json" },
			body: JSON.stringify({ id }),
		});
		const data = await response.json();
		this.props.onDeleteUser([data]);
	};

	handleClick = () => {
		const apiUrl = Endpoints.editUser(this.props.id);

		this.fetchDeleteData(apiUrl, this.props.id);
	};

	render() {
		return (
			<>
				<button
					className="btn"
					onClick={() => this.handleClick()}
					type="button"
				>
					Delete user
				</button>
			</>
		);
	}
}
