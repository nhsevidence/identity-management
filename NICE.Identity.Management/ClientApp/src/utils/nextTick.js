/**
 *  Returns a new promise that resolves after a 0ms setTimeout, effectively clearing the call stack
 */
export const nextTick = async () => {
	return new Promise(resolve => {
		setTimeout(resolve, 0);
	});
};