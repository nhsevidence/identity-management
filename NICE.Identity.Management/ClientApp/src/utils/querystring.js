import queryString from "query-string";

export function queryStringToObject(qryString) {
	return queryString.parse(qryString);
}

export const removeQueryParameter = (url, parameter, value = null) => {
	const urlParts = url.split("?");

	if (urlParts.length < 2) return url;

	let urlParams = urlParts[1].split(/[&;]/g);

	for (let i = urlParams.length; i-- > 0; ) {
		const parParts = urlParams[i].split("=");

		if (
			parParts[0].toLowerCase() === parameter.toLowerCase() &&
			(!value ||
				(value && value.toLowerCase() === (parParts[1] || "").toLowerCase()))
		) {
			urlParams.splice(i, 1);
		}
	}

	const queryString = urlParams.length > 0 ? "?" + urlParams.join("&") : "";
	return urlParts[0] + queryString;
};

export const stripMultipleQueries = (path, queries) => {
	let strippedPath = path;

	queries.forEach((query) => {
		strippedPath = removeQueryParameter(strippedPath, query);
	});

	return strippedPath;
};

export const appendQueryParameter = (url, parameter, value) =>
	`${url}${url.indexOf("?") === -1 ? "?" : "&"}${parameter}=${value}`;
