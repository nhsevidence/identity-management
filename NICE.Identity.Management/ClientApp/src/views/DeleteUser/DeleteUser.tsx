import React, { Component } from "react";
import { RouteComponentProps, Link } from "react-router-dom";

import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";
import { Button } from "@nice-digital/nds-button";
import { PageHeader } from "@nice-digital/nds-page-header";

import { DeleteUserConfirmation } from "../DeleteUserConfirmation/DeleteUserConfirmation";
import { UserType } from "../../models/types";
import { Endpoints } from "../../data/endpoints";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";

import styles from "./DeleteUser.module.scss";
import { Grid, GridItem } from "@nice-digital/nds-grid";

type TParams = { id: string };

type DeleteUserProps = {} & RouteComponentProps<TParams>;

type DeleteUserState = {
	isLoading: boolean;
	isDeleteButtonLoading: boolean;
	error?: Error;
	hasBeenDeleted: boolean;
	data: UserType;
};

export class DeleteUser extends Component<DeleteUserProps, DeleteUserState> {
	constructor(props: DeleteUserProps) {
		super(props);

		this.state = {
			isLoading: true,
			isDeleteButtonLoading: false,
			hasBeenDeleted: false,
			data: {} as UserType,
		};
	}

	handleError = (error: Error) => {
		this.setState({ error });
	};

	fetchData = async (url: string) => {
		this.setState({ isLoading: true });

		let response, data;
		try {
			response = await fetch(url);
			data = await response.json();
		} catch (err) {
			let error: Error = err;

			this.setState({ error });
			return;
		}

		this.setState({ isLoading: false });

		if (response.status === 200) {
			this.setState({ data });
		} else {
			this.setState({ error: new Error(data.message) });
		}
	};

	fetchDeleteData = async () => {
		this.setState({ isDeleteButtonLoading: true });

		let response, data;
		try {
			response = await fetch(Endpoints.user(this.props.match.params.id), {
				method: "DELETE",
				headers: { "Content-Type": "application/json" },
				body: JSON.stringify({ id: this.props.match.params.id }),
			});
			data = await response.json();
		} catch (err) {
			let error: Error = err;

			this.handleError(error);
			return;
		}

		this.setState({ isDeleteButtonLoading: false });

		if (response.status === 200) {
			this.setState({ hasBeenDeleted: true });
		} else {
			this.setState({ error: new Error(data.message) });
		}
	};

	componentDidMount() {
		this.fetchData(Endpoints.user(this.props.match.params.id));
	}

	render() {
		const {
			isLoading,
			isDeleteButtonLoading,
			error,
			hasBeenDeleted,
			data,
		} = this.state;
		const { id } = this.props.match.params;

		let lastBreadcrumb = `${data.first_name} ${data.last_name}`;

		if (isLoading) {
			lastBreadcrumb = "Loading user details";
		}

		if (error) {
			lastBreadcrumb = "";
		}

		return (
			<>
				{hasBeenDeleted ? (
					<DeleteUserConfirmation
						full_name={`${data.first_name} ${data.last_name}`}
					/>
				) : (
					<>
						<Breadcrumbs>
							<Breadcrumb to="/users" elementType={Link}>
								Users
							</Breadcrumb>
							{error ? (
								<Breadcrumb>Error</Breadcrumb>
							) : (
								<Breadcrumb to={`/users/${id}`} elementType={Link}>
									{lastBreadcrumb}
								</Breadcrumb>
							)}
							<Breadcrumb>Delete user</Breadcrumb>
						</Breadcrumbs>

						{!error ? (
							<>
								{isLoading ? (
									<>
										<PageHeader preheading="Confirm" heading="Delete user" />
										<p>Loading...</p>
									</>
								) : (
									<>
										<PageHeader
											preheading="Are you sure you want to delete user?"
											heading={`${data.first_name} ${data.last_name}`}
										/>

										<Grid className="pv--d">
											<GridItem cols={3}>
												<span className={styles.detailsLabel}>
													Email address
												</span>
											</GridItem>
											<GridItem cols={9}>
												<span>{data.email_address}</span>
											</GridItem>
										</Grid>

										<Button
											onClick={this.fetchDeleteData}
											disabled={isDeleteButtonLoading}
										>
											{isDeleteButtonLoading ? "Loading..." : "Confirm"}
										</Button>
										<Button
											to={`/users/${id}`}
											variant="secondary"
											elementType={Link}
											disabled={isDeleteButtonLoading}
										>
											Cancel
										</Button>
									</>
								)}
							</>
						) : (
							<>
								<PageHeader heading="Error" />
								<ErrorMessage error={error}></ErrorMessage>
							</>
						)}
					</>
				)}
			</>
		);
	}
}
