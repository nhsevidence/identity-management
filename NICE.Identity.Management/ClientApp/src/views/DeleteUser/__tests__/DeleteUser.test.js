import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import fetchMock from "fetch-mock";
import toJson from "enzyme-to-json";

import { nextTick } from "../../../utils/nextTick";
import singleUser from "./singleUser.json";
import { DeleteUser } from "../DeleteUser";

describe("DeleteUser", () => {
	const match = {
		params: { id: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	afterEach(fetchMock.reset);

	it("should show loading message before data has been loaded", () => {
		fetchMock.get("*", {});
		const wrapper = shallow(<DeleteUser match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	it("should call fetchData during componentDidMount", () => {
		fetchMock.get("*", {});
		const wrapper = shallow(<DeleteUser match={match} />);
		const instance = wrapper.instance();
		jest.spyOn(instance, "fetchData");
		instance.componentDidMount();
		expect(instance.fetchData).toHaveBeenCalledTimes(1);
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetchMock.get("*", singleUser);
		const wrapper = shallow(<DeleteUser match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetchData function returns 401 error", async () => {
		fetchMock.get("*", 401);
		const wrapper = shallow(<DeleteUser match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetchData function returns 500 error", async () => {
		fetchMock.get("*", 500);
		const wrapper = shallow(<DeleteUser match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when fetchDeleteData fails", async () => {
		const error = new Error("Not allowed");
		fetchMock.get("*", {});
		fetchMock.delete("*", { throws: error });

		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		const instance = wrapper.find(DeleteUser).instance();
		jest.spyOn(instance, "handleError");
		wrapper.find("button").simulate("click");
		await nextTick();
		expect(instance.handleError).toHaveBeenCalledTimes(1);
		expect(instance.handleError).toHaveBeenCalledWith(error);
	});

	it("should show error message when fetchDeleteData returns non-200 error", async () => {
		const serverErrorMessage = "Not authorized";
		fetchMock.get("*", {});
		fetchMock.delete("*", {
			body: { message: serverErrorMessage },
			status: 401,
		});
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		const instance = wrapper.find(DeleteUser).instance();
		jest.spyOn(instance, "handleError");
		wrapper.find("button").simulate("click");
		await nextTick();
		expect(instance.handleError).toHaveBeenCalledTimes(1);
		expect(instance.handleError).toHaveBeenCalledWith(
			new Error(serverErrorMessage),
		);
	});

	it("should disable delete button when clicked", async () => {
		fetchMock.mock("*", singleUser);
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("button").simulate("click");
		expect(wrapper.find("button").props().disabled).toEqual(true);
		expect(wrapper.find("button").text()).toEqual("Loading...");
	});

	it("should display confirmation message once fetchDeleteData is successfully complete", async () => {
		fetchMock.mock("*", {});
		const wrapper = mount(
			<MemoryRouter>
				<DeleteUser match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		wrapper.find("button").simulate("click");
		await nextTick();
		wrapper.update();
		expect(wrapper.find("h1").text()).toEqual("User deleted");
	});
});
