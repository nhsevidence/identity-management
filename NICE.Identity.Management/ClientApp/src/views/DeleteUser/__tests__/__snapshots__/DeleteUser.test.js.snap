// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DeleteUser should match the snapshot after data has been loaded 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={[Function]}
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType={[Function]}
      to="/users/1"
    >
      Vernita Green
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Delete user
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="Vernita Green"
    preheading="Are you sure you want to delete user?"
  />
  <Grid
    className="pv--d"
    elementType="div"
    gutter="standard"
    horizontalAlignment="left"
    reverse={false}
    verticalAlignment="top"
  >
    <GridItem
      cols={3}
      elementType="div"
    >
      <span
        className="detailsLabel"
      >
        Email address
      </span>
    </GridItem>
    <GridItem
      cols={9}
      elementType="div"
    >
      <span>
        vernita.green@gmail.com
      </span>
    </GridItem>
  </Grid>
  <Button
    disabled={false}
    onClick={[Function]}
    variant="primary"
  >
    Confirm
  </Button>
  <Button
    disabled={false}
    elementType={[Function]}
    to="/users/1"
    variant="secondary"
  >
    Cancel
  </Button>
</Fragment>
`;

exports[`DeleteUser should show error message when fetchData function returns 401 error 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={[Function]}
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Error
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Delete user
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="Error"
  />
  <ErrorMessage
    error={[FetchError: invalid json response body at http://localhost:3001/users/1 reason: Unexpected end of JSON input]}
  />
</Fragment>
`;

exports[`DeleteUser should show error message when fetchData function returns 500 error 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={[Function]}
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Error
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Delete user
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="Error"
  />
  <ErrorMessage
    error={[FetchError: invalid json response body at http://localhost:3001/users/1 reason: Unexpected end of JSON input]}
  />
</Fragment>
`;
